### PR TITLE
GenotypeConcordance can output an annotated VCF.

### PR DESCRIPTION
**Pull request by nh13**
I find it useful to manually inspect individual variant calls based on their concordance state using this tool.  I also find it useful for filtering for variants with a particular concordance state (ex. FP).

It should also be possible (not implemented here) to re-compute the genotype concordance metrics using the new output.  This would be useful if you want to remove variants based on some criteria (ex. low-frequency, info/format tag/value) and re-compute the metrics.  Of course, one could always filter the original input VCF that is given to `GenotypeConcordance`.

(@yfarjoun `||` @tfenne `||` @droazen) would you mind taking a look?

### CODE REVIEW

**Comment by tfenne** [comment]
@nh13 This is fantastic and I'm excited to have this available.  I can probably review over the weekend.  Would be good for someone else to take a peek too probably though?

**Comment by nh13** [comment]
@droazen or @yfarjoun would you also care to take a look?  My two concerns wit this PR are (1) there is not a really good way to merge VCF headers, and (2) some of the methods I use to re-compute the concordance state for output to the VCF could be changed so that I am not re-computing.

**Comment by yfarjoun** [comment]
This is a very nice thing to have @kbergin you would find use for this quite soon I think...care to review it?

**Comment by nh13** [comment]
Just as a head's up, the second commit caused me to refactor a lot of code, since I hit the case where the reference allele for two different variant contexts at the same side were different.  Therefore, I needed to normalize them, and that cause a domino effect in terms of needed to get access to certain internals, such as the normalized alleles for the truth and call variant contexts.  I also have increased my profound hate for the vcf code in htsjdk.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 1]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
                 unClassifiedStatesMap.put(condition, count);
             }
 
+            // write out an annotated VCF
+            if (writer != null) {
+                final VariantContextBuilder builder;
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I think this could move down into your `if` now?

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 2]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
...
+                    truthContext = tuple.leftVariantContext.get();
+                }
+                if (tuple.rightVariantContext.isPresent()) {
+                    callContext = tuple.rightVariantContext.get();
+                }
+
+                // Get the alleles for each genotype
+                final Alleles alleles = normalizeAlleles(truthContext, TRUTH_SAMPLE, callContext, CALL_SAMPLE);
+
+                // Ignore if both are one of mixed, filtered, or missing.
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Is this comment still accurate?  I.e. alleles will be empty if both genotypes are mixed/filtered/missing?  If so, maybe reword to make that clear?

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 3]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
...
+                if (!alleles.allAlleles.isEmpty()) {
+                    final List<Allele> allAlleles   = alleles.asAlleles();
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: `asAlleles` -> `asList`

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 4]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
...
+                        genotypeBuilder.name(outputTruthSampleName);
+                        genotypeBuilder.alleles(truthAlleles);
+                        if (!truthGenotype.hasAnyAttribute(VCFConstants.GENOTYPE_KEY)) {
+                            genotypeBuilder.attribute(VCFConstants.GENOTYPE_KEY, MISSING_VALUE_v4);
+                        }
+                        genotypes.add(genotypeBuilder.make());
+                    }
+                    else {
+                        final GenotypeBuilder genotypeBuilder = new GenotypeBuilder(outputTruthSampleName);
+                        genotypeBuilder.alleles(Arrays.asList(Allele.NO_CALL, Allele.NO_CALL));
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I think here, if `MISSING_SITES_HOM_REF == true` then it would be more helpful to produce an output genotype with ref/ref instead.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 5]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
...
+                        genotypeBuilder.attribute(VCFConstants.GENOTYPE_KEY, MISSING_VALUE_v4);
+                        genotypes.add(genotypeBuilder.make());
+                    }
+                    if (callContext != null && !callAlleles.isEmpty()) {
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: This whole `if` block is repetitious.  Could be extracted to a method that takes in the sample name, genotype and set of alleles...

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 6]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
...
+                        genotypeBuilder.attribute(VCFConstants.GENOTYPE_KEY, MISSING_VALUE_v4);
+                        genotypes.add(genotypeBuilder.make());
+                    }
+
+                    // set the alleles and genotypes
+                    builder.alleles(alleles.allAlleles).genotypes(genotypes);
+
+                    // set the concordance state attribute
+                    final TruthAndCallStates state = GenotypeConcordance.determineState(truthContext, TRUTH_SAMPLE, callContext, CALL_SAMPLE, MIN_GQ, MIN_DP);
+                    final ContingencyState[] stateArray = scheme.getConcordanceStateArray(state.truthState, state.callState);
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I think it would be nicer to output something other than the contingency values here because I think those are a bit confusing until you understand their purpose.  I.e. seeing something labelled `TP,FP` is a little odd.  But I don't think there's really much else that could be emitted instead that's ready to go.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 7]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
+    /** A simple structure to return the results of getAlleles.  NB: truthAllele1/2 and callAllele1/2 may be null if
+     * no first or second allele was present respectively. */
+    private static class Alleles {
+        public final OrderedSet<String> allAlleles;
+        public final String truthAllele1;
+        public final String truthAllele2;
+        public final String callAllele1;
+        public final String callAllele2;
+
+
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Extra newline.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 8]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
+            this.truthAllele1 = (truthAllele1 == null) ? null : this.allAlleles.get(allAlleles.indexOf(truthAllele1));
+            this.truthAllele2 = (truthAllele2 == null) ? null : this.allAlleles.get(allAlleles.indexOf(truthAllele2));
+            this.callAllele1  = (callAllele1 == null) ? null : this.allAlleles.get(allAlleles.indexOf(callAllele1));
+            this.callAllele2  = (callAllele2 == null) ? null : this.allAlleles.get(allAlleles.indexOf(callAllele2));
+        }
 
-        // TODO: what about getPloidy()
 
-        Genotype truthGenotype = null, callGenotype = null;
+        public List<Allele> asAlleles() {
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: `asList`

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 9]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
+            if (allAlleles.isEmpty()) {
+                return Collections.emptyList();
+            }
+            else {
+                final List<Allele> alleles = new ArrayList<>(4);
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Either remove `4` or change to `this.allAlleles.size()` please.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 10]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
+                int idx = 0;
+                while (idx < this.allAlleles.size()) {
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: why not just a regular for loop?

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 11]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
-            // Genotype level checks
-            truthGenotype = truthContext.getGenotype(truthSample);
-            if (truthGenotype.isNoCall())           truthState = TruthState.NO_CALL;
-            else if (truthGenotype.isFiltered())    truthState = TruthState.GT_FILTERED;
-            else if ((truthGenotype.getGQ() != -1) && (truthGenotype.getGQ() < minGq)) truthState = TruthState.LOW_GQ;
-            else if ((truthGenotype.getDP() != -1) && (truthGenotype.getDP() < minDp)) truthState = TruthState.LOW_DP;
-            // Note.  Genotype.isMixed means that it is called on one chromosome and NOT on the other
-            else if ((truthGenotype.isMixed())) truthState = TruthState.NO_CALL;
+        public List<Allele> truthAlleles() {
+            if (allAlleles.isEmpty() || this.truthAllele1 == null) {
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: It feels a little odd that in the constructor you check `truthAllele2` for null independently, but here you seem to assume it must be non-null if `truthAllele1` is.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 12]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
-            // Genotype level checks
-            callGenotype = callContext.getGenotype(callSample);
-            if (callGenotype.isNoCall())           callState = CallState.NO_CALL;
-            else if (callGenotype.isFiltered())    callState = CallState.GT_FILTERED;
-            else if ((callGenotype.getGQ() != -1) && (callGenotype.getGQ() < minGq)) callState = CallState.LOW_GQ;
-            else if ((callGenotype.getDP() != -1) && (callGenotype.getDP() < minDp)) callState = CallState.LOW_DP;
-                // Note.  Genotype.isMixed means that it is called on one chromosome and NOT on the other
-            else if ((callGenotype.isMixed())) callState = CallState.NO_CALL;
+        public List<Allele> callAlleles() {
+            if (allAlleles.isEmpty() || this.callAllele1 == null) {
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Same here about the null-checking.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 13]
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
...
+    /** Gets the alleles for the truth and call genotypes.  In particular, this handles the case where indels can have different
+     * reference alleles. */
+    final private static Alleles normalizeAlleles(final VariantContext truthContext,
+                                                  final String truthSample,
+                                                  final VariantContext callContext,
+                                                  final String callSample) {
+
+        final Genotype truthGenotype, callGenotype;
+
+        if (truthContext == null || truthContext.isMixed() || truthContext.isFiltered()) truthGenotype = null;
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I'm not sure if it's done elsewhere, but before calling `isMixed()` would it make sense to produce sub-contexts that are just for the sample of interest?  I.e. we don't much care if a second sample has an allele that makes the context mixed.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 14]
```diff
@@ -546,19 +692,15 @@ final public static TruthAndCallStates determineState(final VariantContext truth
                 // Truth reference is shorter than call reference
                 final String suffix = getStringSuffix(callRef, truthRef, "Ref alleles mismatch between: " + truthContext + " and " + callContext);
                 truthRef = truthRef + suffix;
-                if (null == truthState) {
-                    truthAllele1 = truthGenotype.getAllele(0).getBaseString() + suffix;
-                    truthAllele2 = truthGenotype.getAllele(1).getBaseString() + suffix;
-                }
+                truthAllele1 = truthGenotype.getAllele(0).getBaseString() + suffix;
+                truthAllele2 = truthGenotype.getAllele(1).getBaseString() + suffix;
```
Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Does this work properly on indels that are not repetitive?  I.e. shouldn't the `suffix` be spliced in between the shorter ref string and the rest of the allele?  My brain hurts thinking about this.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 15]
```diff
@@ -109,6 +108,13 @@
             "<li>TN - True negatives are correctly called reference sites</li>" +
             "<li>Validated genotypes - are TP sites where the exact genotype (HET or HOM-VAR) has been validated </li> "     +
             "</ul>"+
+            "" +
+            "<h4>VCF Output:</h4>" +
+            "<ul>" +
+            "<li>The concordance state will be stored in the \"ST\" tag in the INFO field.</li>" +
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Will use `CONC_ST`

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 16]
```diff
@@ -109,6 +108,13 @@
...
+            "<li>If the truth and call sample names match, the call will have the \"_CALL\" suffix and the truth will have the " +
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Good idea, that makes it much simpler.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 17]
```diff
@@ -118,9 +124,12 @@
     public File CALL_VCF;
 
     @Option(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "Basename for the two metrics files that are to be written." +
-            " Resulting files will be <OUTPUT>" + SUMMARY_METRICS_FILE_EXTENSION + "  and <OUTPUT>" + DETAILED_METRICS_FILE_EXTENSION + ".")
+            " Resulting files will be <OUTPUT>" + SUMMARY_METRICS_FILE_EXTENSION + " and <OUTPUT>" + DETAILED_METRICS_FILE_EXTENSION + ".")
     public File OUTPUT;
 
+    @Option(doc = "Output VCF annotated with concordance information.", optional = true)
+    public File OUTPUT_VCF = null;
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Will do.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 18]
```diff
@@ -118,9 +124,12 @@
...
+
     @Option(shortName = "TS", doc="The name of the truth sample within the truth VCF")
     public String TRUTH_SAMPLE;
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Not hard.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 19]
```diff
@@ -163,6 +172,9 @@
     protected GenotypeConcordanceCounts indelCounter;
     public GenotypeConcordanceCounts getIndelCounter() { return indelCounter; }
 
+    public static final String ContingencyStateTag = "ST";
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: +1 scala

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 20]
```diff
@@ -163,6 +172,9 @@
...
+    public static final VCFHeaderLine ContingencyStateHeaderLine = new VCFInfoHeaderLine(ContingencyStateTag, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, "The genotype concordance contingency state(s)");
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 21]
```diff
@@ -163,6 +172,9 @@
...
+
     // TODO: add optimization if the samples are in the same file
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I don't have a need for it so I will defer to another PR.

Comment (tfenne) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I was suggesting deleting the three comment lines - not implementing them!

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 22]
```diff
@@ -260,6 +276,30 @@ private boolean indexExists(final File vcf) {
         // Verify that both VCFs have the same Sequence Dictionary
         SequenceUtil.assertSequenceDictionariesEqual(truthReader.getFileHeader().getSequenceDictionary(), callReader.getFileHeader().getSequenceDictionary());
 
+        VariantContextWriter writer = null;
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 23]
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
                 unClassifiedStatesMap.put(condition, count);
             }
 
+            // write out an annotated VCF
+            if (writer != null) {
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 1 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 2 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: It is, will do.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 3 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 4 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 5 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 6 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I agree the output could be improved, lets think about it and add it in later.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 7 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Just making sure you were checking ;)

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 8 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 9 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Done.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 10 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Why indeed.  Fixed.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 11 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I'll add a check in the constructor.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 12 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Fixed.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 13 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: The `PairedVariantSubContextIterator` creates a single-sample sub-context.  Yes, it's that confusing.

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 14 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: I think the current implementation works for a deletion, but not an insertion.  It should be added in-before the rest of the allele, which for deletions is empty, but for an insertion could be many bases.  I'll see if I can fix it.  Hrmn.

**Comment by nh13** [comment]
@tfenne, thanks for the comments, I implemented almost all of what you suggested.  @kbergin if possible, could you take a look?

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 16 above)
Comment (kbergin) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: 👍 that thought

Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: @kbergin it's in a more recent commit, can you work off of that for your review?

**Comment by kbergin** [comment]
@nh13 I have looked through this today and I think it all makes sense, but I'd like to look at it again on Monday when my brain has processed it more, it's a big change in code that always turns my head around! 

In the mean time, GenotypeConcordance tests are failing

**Comment by nh13** [comment]
@kbergin thanks for taking a look.  I have modified the tests too many times after folks' feedback, and it's painful to do, so I am going to wait until you have reviewed before I fix them up one last time (the output files differ, which is expected).  Thanks!

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 16 above)
Comment (kbergin) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Oh I did, I was just reading through what was commented on before your more recent changes out of curiousity, and wanted to "like" the change lol

**Comment by kbergin** [comment]
@nh13 That makes a lot of sense!!!!

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 24]
```diff
@@ -341,9 +375,119 @@ private boolean indexExists(final File vcf) {
...
+    }
+
+    private void addToGenotypes(final List<Genotype> genotypes,
+                                final VariantContext ctx,
+                                final String inputSampleName,
+                                final String outputSampleName,
+                                final List<Allele> allAlleles,
+                                final List<Allele> ctxAlleles,
+                                final boolean missingSitesHomRef) {
+        if (ctx != null && !ctxAlleles.isEmpty()) {
```
Comment (kbergin) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: A comment explaining this function a little might be useful for future readers.

**Comment by nh13** [comment]
@kbergin just checking in to see if you have finished your review.

**Comment by kbergin** [comment]
@nh13 Sorry, I was trying to follow the new github review structure, it seems now you 'start a review' and all the comments are pending until you mark it as a 'finished review' and then it posts it. So yup, all set! Can you see my couple comments from the other day?

File: src/main/java/picard/vcf/GenotypeConcordance.java [hunk 25]
```diff
@@ -341,9 +375,119 @@ private boolean indexExists(final File vcf) {
...
+        }
+
+        // Get the alleles for each genotype.  No alleles will be extracted for a genotype if the genotype is
+        // mixed, filtered, or missing.
+        final Alleles alleles = normalizeAlleles(truthContext, TRUTH_SAMPLE, callContext, CALL_SAMPLE);
+
+        // There will be no alleles if both genotypes are one of mixed, filtered, or missing.  Do not output any
+        // variant context in this case.
+        if (!alleles.allAlleles.isEmpty()) {
+
```
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Fixed

File: src/main/java/picard/vcf/GenotypeConcordance.java (hunk 24 above)
Comment (nh13) [inline src/main/java/picard/vcf/GenotypeConcordance.java]: Fixed.

**Comment by nh13** [comment]
@tfenne and @kbergin  I have fixed up the commits so I think it is ready for merge after tests pass, except....

...@yfarjoun `ByIntervalListVariantContextIterator` could return a variant multiple times if the variant overlaps multiple intervals.  I fixed that, and added tests (there were none).  Care to take a look at the first commit?   If you give me 👍 I can merge when tests pass.

**Comment by kbergin** [comment]
Yay! 👍

File: src/main/java/picard/vcf/ByIntervalListVariantContextIterator.java [hunk 26]
```diff
@@ -20,33 +23,58 @@
...
         advance();
-        return this.currentIterator.next();
+        return ctx;
     }
 
     @Override public void remove() {
         throw new UnsupportedOperationException();
     }
 }
+
```
Comment (yfarjoun) [inline src/main/java/picard/vcf/ByIntervalListVariantContextIterator.java]: two extra newlines?

File: src/main/java/picard/vcf/ByIntervalListVariantContextIterator.java [hunk 27]
```diff
@@ -20,33 +23,58 @@
...
+        this.reader    = reader;
         this.intervals = intervals.uniqued().iterator();
+        this.advance();
+    }
+
+    /** Returns true if the variant and interval overlap. */
+    private boolean overlapsInterval(final VariantContext ctx, final Interval interval) {
+        if (!ctx.getContig().equals(interval.getContig())) return false;
+        else if (CoordMath.overlaps(ctx.getStart(), ctx.getEnd(), interval.getStart(), interval.getEnd())) return true;
+        else if (CoordMath.encloses(ctx.getStart(), ctx.getEnd(), interval.getStart(), interval.getEnd())) return true;
```
Comment (nh13) [inline src/main/java/picard/vcf/ByIntervalListVariantContextIterator.java]: Done.

**Comment by nh13** [comment]
Waiting for tests to pass.
